### PR TITLE
feat(examples): add p/demo/avl/list

### PR DESCRIPTION
**Pull request by moul**
Could become the preferred method for managing top-level lists. 

It can likely be utilized extensively in the existing codebase, potentially for most instances involving a top-level slice or those involving an avl.Tree using a seqid.

### CODE REVIEW

**Comment by Gno2D2** [comment]
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **SKIP**: Do not block the CI for this PR


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 The pull request head branch must be up-to-date with its base ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/keeping-your-pull-request-in-sync-with-the-base-branch))


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 The pull request was created from a fork (head branch repo: moul/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>The pull request head branch must be up-to-date with its base (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Head branch (moul:dev/moul/avllist) is up to date with base (master): behind by 0 / ahead by 2

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**SKIP**: Do not block the CI for this PR</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

File: examples/gno.land/p/demo/avl/list/list.gno [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+// Clear removes all elements from the list.
+//
+// Example:
+//
+//	l := list.New()
+//	l.Append(1, 2, 3)
+//	l.Clear()
+//	println(l.Len()) // Output: 0
+func (l *List) Clear() {
```
[the remark this answers is not in the document]
Comment (moul) [reply examples/gno.land/p/demo/avl/list/list.gno]: b482ec125

File: examples/gno.land/p/demo/avl/list/list.gno [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+// Example:
+//
+//	l := list.New()
+//	l.Append(1, 2, 3, 4, 5)
+//
+//	println(l.Range(1, 4))   // Output: [2 3 4]
+//	println(l.Range(-1, 2))  // Output: [1 2]
+//	println(l.Range(3, 999)) // Output: [4 5]
+//	println(l.Range(3, 2))   // Output: nil
+func (l *List) Range(startIndex, endIndex int) []interface{} {
```
[the remark this answers is not in the document]
Comment (moul) [reply examples/gno.land/p/demo/avl/list/list.gno]: b482ec125

File: examples/gno.land/p/demo/avl/list/list.gno [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+}
+
+// New creates a new empty list.
+//
+// Example:
+//
+//	l := list.New()
+//	l.Append(1, 2, 3)
+//	println(l.Len()) // Output: 3
+func New() *List {
```
[the remark this answers is not in the document]
Comment (moul) [reply examples/gno.land/p/demo/avl/list/list.gno]: b482ec125

File: examples/gno.land/p/demo/avl/list/list.gno [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+// Example:
+//
+//	l := list.Make(3, "default")
+//	println(l.Get(0)) // Output: "default"
+//	println(l.Len())  // Output: 3
+//
+//	// With nil default value
+//	l2 := list.Make(2, nil)
+//	println(l2.Get(0)) // Output: nil
+func Make(length int, defaultValue interface{}) *List {
```
[the remark this answers is not in the document]
Comment (moul) [reply examples/gno.land/p/demo/avl/list/list.gno]: b482ec125

File: examples/gno.land/p/demo/avl/list/list.gno [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+
+	// Then iterate through them in correct order
+	for i := 0; i < size; i++ {
+		if fn(i, values[i]) {
+			return
+		}
+	}
+}
+
+// Clone creates a deep copy of the list.
```
[the remark this answers is not in the document]
Comment (moul) [reply examples/gno.land/p/demo/avl/list/list.gno]: 7e2a01253

File: examples/gno.land/p/demo/avl/list/list.gno [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+		endIndex = size
+	}
+	if startIndex >= endIndex {
+		return nil
+	}
+
+	count := endIndex - startIndex
+	result := make([]interface{}, count)
+
+	// Fill the slice from end to start to maintain correct order
```
Comment (thehowl) [inline examples/gno.land/p/demo/avl/list/list.gno:206]: Why is this reversed?

File: examples/gno.land/p/demo/avl/list/list.gno [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+	size := l.tree.Size()
+	values := make([]interface{}, size)
+
+	// First collect values in reverse order
+	i := size - 1
+	l.tree.IterateByOffset(0, size, func(_ string, value interface{}) bool {
+		values[i] = value
+		i--
+		return false
+	})
```
Comment (thehowl) [inline examples/gno.land/p/demo/avl/list/list.gno:249]: If reverse here is necessary, you can use ReverseIterateByOffset and call fn() directly

File: examples/gno.land/p/demo/avl/list/list.gno [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+	values := make([]interface{}, size)
+	i := size - 1
+	l.tree.IterateByOffset(0, size, func(_ string, value interface{}) bool {
+		values[i] = value
+		i--
+		return false
+	})
+
+	// Add values to new list in correct order
+	for _, value := range values {
```
Comment (thehowl) [inline examples/gno.land/p/demo/avl/list/list.gno:292]: similarly here

**Review by thehowl — APPROVED** [review]

**Review by thehowl — APPROVED** [review]
I figured out why the avl trees are reversed in IterateByOffset. It's a bug.

I'd say just use the Reverse functions for now, I'll make a PR to fix avl.

**Pull request merged**